### PR TITLE
perf(lsp): use faster version of str_byteindex

### PR DESCRIPTION
**Pull request by lewis6991**
perf(lsp): optimize input validation in str_byteindex

The semantic tokens response handler contains a hot loop which makes
lots of calls to vim.str_byteindex which takes most of the time of the
entire handler.

Optimized input validation in vim.str_byteindex. Should be roughly
4x faster.

    name                              total    count      avg
    ────────────────────────────────────────────────
    vim.str_byteindex (original)   635.67ms   263048   2.42us
    vim.str_byteindex (new)        159.20ms   263016   0.61us
    vim._str_byteindex             190.01ms   786164   0.24us

### CODE REVIEW

**Comment by lewis6991** [comment]
@tris203

**Comment by tris203** [comment]
Seems sensible 

I will rebench the lua implementation of this function with the added guard

**Comment by lewis6991** [comment]
> I will rebench the lua implementation of this function with the added guard

I just tried the Lua version and it was around 10-15% faster. Not sure if it's worth going with.

I also tried using the FFI but that was around the same.

I've guarded some more validation and removed the specialised version. This should be good enough for now.

File: runtime/lua/vim/_editor.lua [hunk 1]
```diff
@@ -764,20 +764,28 @@ function vim.str_byteindex(s, encoding, index, strict_indexing)
     return vim._str_byteindex(s, old_index, use_utf16) or error('index out of range')
   end
 
-  vim.validate('s', s, 'string')
-  vim.validate('index', index, 'number')
+  -- Note we bypass vim.validate for performance reasons
+
+  if type(s) ~= 'string' or type(index) ~= 'number' then
```
Comment (justinmk) [inline runtime/lua/vim/_editor.lua:769]:
```suggestion
  -- Avoid vim.validate for performance.
  if type(s) ~= 'string' or type(index) ~= 'number' then
```

**Review by justinmk — APPROVED** [review]

**Pull request merged**